sunpy/map/maputils.py

```diff
@@ -164,11 +166,73 @@ def solar_angular_radius(coordinates):
     return sun._angular_radius(coordinates.rsun, coordinates.observer.radius)
 
 
-def sample_at_coords(smap, coordinates):
+def get_neighboring_indices(indices, position, dimensions):
     """
-    Samples the data in a map at given series of coordinates.
-    Uses nearest-neighbor interpolation of coordinates in map, as
-    it effectively uses array indexing.
+    Get the neighboring indices relative to given indices.
+
+    Parameters
+    ----------
+    indices : 2-D ndarray of integers with shape (2, n)
+        Indices to find neighbors for.
+    position : 1-D ndarray of shape (2,) or 2-element list
+        x and y position of neighbor relative to given indices.
+    dimensions : 1-D ndarray of shape (2,) or 2-element list
+        x and y dimensions of the map.
+
+    Returns
+    -------
+    neighbor_indices : 2-D ndarray of integers with shape (2, ...)
+        Neighboring indices. Those out of bound of the map are masked
+        out.
+
+    Examples
+    --------
+    >>> indices = np.array([[0, 1, 2, 3, 4], [0, 1, 2, 3, 4]])
+    >>> indices
+    array([[0, 1, 2, 3, 4],
+           [0, 1, 2, 3, 4]])
+
+    >>> # Right neighbors
+    >>> get_neighboring_indices(indices, [1, 0], [5, 5])
+    array([[1, 2, 3, 4],
+           [0, 1, 2, 3]])
+
+    >>> # Left neighbors
+    >>> get_neighboring_indices(indices, [-1, 0], [5, 5])
+    array([[0, 1, 2, 3],
+           [1, 2, 3, 4]])
+
+    >>> # Top neighbors
+    >>> get_neighboring_indices(indices, [0, 1], [5, 5])
+    array([[0, 1, 2, 3],
+           [1, 2, 3, 4]])
+
+    >>> # Bottom neighbors
+    >>> get_neighboring_indices(indices, [0, -1], [5, 5])
+    array([[1, 2, 3, 4],
+           [0, 1, 2, 3]])
+
+    """
+    x, y = indices + np.array(position).reshape(2, 1)
+    in_bound = (
+        (x >= 0)
+        & (x <= dimensions[0] - 1)
+        & (y >= 0)
+        & (y <= dimensions[1] - 1)
+    )
+    return np.vstack((x[in_bound], y[in_bound]))
+
+
+def sample_at_coords(
+    smap,
+    coordinates,
+    method="astropy",
+    neighbor_positions=[[1, 0], [-1, 0], [0, 1], [0, -1]],
```

Comment:
See my other comment about this API being awkward.  Also, why are you not including diagonals?

Reply:
It's just to make the resulting area (red dots in original comment) surrounding the curve (blue dots) a little smoother. Do we want to include the diagonal neighbors (North-East: `[1, 1]`, North-West: `[-1, 1]`, South-East: `[1, -1]`, South-West: `[-1, -1]`) by default? Or maybe we can expose two arguments: `radius: int` and `include_diagonal_neighbors: bool`? I'm not married to the name of the second argument, [scipy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.generate_binary_structure.html#scipy.ndimage.generate_binary_structure) calls it `connectivity`, so we can call it that, too. But since we're dealing with a 2D image, `connectivity` can only be '1' (N, S, E, W) or '2' (also including NE, NW, SE, SW).

Reply:
The `connectivity` thing is too complicated. I decided to just pick points within a pixel `radius`, which is the shortest path from one pixel to another. This will also include the diagonals.